filter_sarif.py

```diff
@@ -1,147 +1,218 @@
     result = True
-    for s, fp, rp in patterns:
-        if match(rp, rule) and match(fp, path):
-            result = s
+
+    LOG.debug("Matching patterns against %s:%s:%s", path, rule, message)
+
+    for sign, file_pattern, rule_pattern, message_pattern in patterns:
+        LOG.debug("Matching against %s", (sign, file_pattern, rule_pattern, message_pattern))
+        if (path is None or file_pattern is None or
+                match(file_pattern, path)) and (rule is None or rule_pattern is None or match(rule_pattern, rule)) and (
+                    message_pattern is None or message is None or re.search(message_pattern, message)):
+            LOG.debug("Matched")
+            result = sign
+
     return result
 
 
-def parse_pattern(line):
-    sepchar = ':'
-    escchar = '\\'
-    file_pattern = ''
-    rule_pattern = ''
-    seen_separator = False
+def parse_pattern(line: str) -> Tuple[bool, str, Optional[str], Optional[str]]:
+    """Parse a pattern line into a tuple of (sign, file_pattern, rule_pattern, message_pattern)."""
     sign = True
+    file_pattern: str = ''
+    rule_pattern: Optional[str] = None
+    message_pattern: Optional[str] = None
+    seen_separator: int = 0
 
     # inclusion or exclusion pattern?
     uline = line
     if line:
-        if line[0] == '-':
+        if line[0] == exclude_char:
             sign = False
             uline = line[1:]
-        elif line[0] == '+':
+        elif line[0] == include_char:
             uline = line[1:]
 
+    # parse patterns character by character
     i = 0
-    while i < len(uline):
+    line_length = len(uline)
+
+    while i < line_length:
         c = uline[i]
-        i = i + 1
+        i += 1
         if c == sepchar:
-            if seen_separator:
-                raise Exception('Invalid pattern: "' + line + '" Contains more than one separator!')
-            seen_separator = True
+            seen_separator += 1
+            if seen_separator > 2:
+                raise Exception(
+                    f'Invalid pattern: "{line}" contains more than two separators. Use {escchar}{sepchar} to escape the separator.'
+                )
+            if seen_separator == 1:
+                rule_pattern = ''
+            elif seen_separator == 2:
+                message_pattern = ''
             continue
         elif c == escchar:
-            nextc = uline[i] if (i < len(uline)) else None
-            if nextc in ['+' , '-', escchar, sepchar]:
-                i = i + 1
+            nextc = uline[i] if (i < line_length) else None
+            if nextc in special_chars:
+                i += 1
                 c = nextc
-        if seen_separator:
-            rule_pattern = rule_pattern + c
+        if seen_separator == 1:
+            rule_pattern = rule_pattern + c if rule_pattern is not None else None
+        elif seen_separator == 2:
+            message_pattern = message_pattern + c if message_pattern is not None else None
```

Comment:
ditto

Reply:
As above, type checking and signalling possible state